### DISCUSSION TITLE
Whats the nvidia blob implementations dri implementation called?

### DISCUSSION THREAD

**Discussion by YusufKhan-gamedev** [NVIDIA/open-gpu-kernel-modules#321] [Q&A] [answered]
Simple question, whats the nvidia blob implementations name configured as, should be the result of `xdriinfo driver 0` or `xdriinfo driver 1`.

**Comment by aaronp24** [comment, marked as answer]
The NVIDIA GLX implementation doesn't support the `glXGetScreenDriver` function that xdriinfo uses, so it just prints
```
Screen "0" is not direct rendering capable.
```
The nvidia-drm.ko kernel module registers itself with the drm subsystem with the name "[nvidia-drm](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/88636e701daad77ae77a0bd861bbb0ac28a31763/kernel-open/nvidia-drm/nvidia-drm-drv.c#L850)".

**Reply by YusufKhan-gamedev** [reply]
Ah, I see, I was hoping to detect the proprietary driver through glXGetScreenDriver()

**Reply by aaronp24** [reply]
If you want to detect the presence of the NVIDIA X driver, you can check for the NV-CONTROL extension. You can check whether each particular X screen is using the NVIDIA driver by querying the GLX vendor string or by using the [XNVCTRLIsNvScreen](https://github.com/NVIDIA/nvidia-settings/blob/732b38ee18a1d9cdee88feab6298dd61552fa900/src/libXNVCtrl/NVCtrlLib.h#L64) function.

**Reply by YusufKhan-gamedev** [reply]
Detecting the presence of the kernel driver is fine _enough_.

**Reply by YusufKhan-gamedev** [reply]
Wait, sorry, in that case where is dri loaded from?

**Reply by aaronp24** [reply]
"dri" can refer to a lot of things. Which component specifically are you asking about? If you're asking about direct OpenGL rendering in general, the NVIDIA GLX driver (and EGL and Vulkan) implement direct rendering on their own without using the X server's DRI2 or DRI3 extensions and without using the kernel's drm system directly.

There are some interop cases such as rendering on an NVIDIA GPU and displaying on another vendor's GPU that do use the kernel's drm subsystem and the DRI3 extension.

**Reply by YusufKhan-gamedev** [reply]
I meant the Xserver's DRI(or Mesa's /lib/dri/*_dri.so), thanks for the clarification!
